src/app/modules/submission/email-submission/email-submission.controller.ts

```diff
@@ -30,267 +32,274 @@ const logger = createLoggerWithLabel(module)
-    replyToEmails: EmailSubmissionService.extractEmailAnswers(parsedResponses),
-    form,
-    submission,
-    attachments,
-    dataCollationData: emailData.dataCollationData,
-    formData: emailData.formData,
-  })
-  if (sendAdminEmailResult.isErr()) {
-    logger.error({
-      message: 'Error sending submission to admin',
-      meta: logMetaWithSubmission,
-      error: sendAdminEmailResult.error,
-    })
-    const { statusCode, errorMessage } = mapRouteError(
-      sendAdminEmailResult.error,
-    )
-    return res.status(statusCode).json({
-      message: errorMessage,
-      spcpSubmissionFailure: false,
-    })
-  }
+        // Save submission to database
+        return EmailSubmissionService.hashSubmission(
+          emailData.formData,
+          attachments,
+        )
+          .andThen((submissionHash) =>
+            EmailSubmissionService.saveSubmissionMetadata(form, submissionHash),
+          )
+          .map((submission) => ({
+            form,
+            parsedResponses,
+            submission,
+            emailData,
+          }))
+          .mapErr((error) => {
+            logger.error({
+              message: 'Error while saving metadata to database',
+              meta: logMeta,
+              error,
+            })
+            return error
+          })
+      })
+      .andThen(({ form, parsedResponses, submission, emailData }) => {
+        const logMetaWithSubmission = {
+          ...logMeta,
+          submissionId: submission._id,
+        }
 
-  // MyInfo access token is single-use, so clear it
-  res.clearCookie(MYINFO_COOKIE_NAME, MYINFO_COOKIE_OPTIONS)
-  // Return the reply early to the submitter
-  res.json({
-    message: 'Form submission successful.',
-    submissionId: submission.id,
-  })
+        logger.info({
+          message: 'Sending admin mail',
+          meta: logMetaWithSubmission,
+        })
 
-  // Send email confirmations
-  return SubmissionService.sendEmailConfirmations({
-    form,
-    parsedResponses,
-    submission,
-    attachments,
-    autoReplyData: emailData.autoReplyData,
-  }).mapErr((error) => {
-    logger.error({
-      message: 'Error while sending email confirmations',
-      meta: logMetaWithSubmission,
-      error,
-    })
-  })
+        // Send response to admin
+        // NOTE: This should short circuit in the event of an error.
+        // This is why sendSubmissionToAdmin is separated from sendEmailConfirmations in 2 blocks
+        return MailService.sendSubmissionToAdmin({
+          replyToEmails: EmailSubmissionService.extractEmailAnswers(
+            parsedResponses,
+          ),
+          form,
+          submission,
+          attachments,
+          dataCollationData: emailData.dataCollationData,
+          formData: emailData.formData,
+        })
+          .map(() => ({
+            form,
+            parsedResponses,
+            submission,
+            emailData,
+            logMeta,
```

Comment:
why do you need to pass `logMeta` here? isn't it in scope throughout the function?

Reply:
each `andThen` is a function. hence, variables created in `andThen` is only in scope for that single `andThen` :(

Reply:
on i see the confusion; this should be `logMetaWithSubmission` - this is bad namign from me @_@ thanks for pointing it out